// resti/examples/node/app/controller.js
var Todo = require('./models/Todo');

exports.list = function(req, res, next) {
	Todo.find(function(err, todos) {
		if (err) return next(err);
		res.json(todos);
	});
};

exports.create = function(req, res, next) {
	var todo = new Todo({
		title: req.body.title
	});

	todo.save(function(err, todo) {
		if (err) {
			return next(err);
		}
		res.json(todo);
	});
};

exports.findOne = function(req, res, next) {
	Todo.findById(req.params.id, function(err, todo) {
		if (err) {
			next(err);
		} else if (todo) {
			req.todo = todo;
			next();
		} else {
			next(new Error('failed to load todo'));
		}
	});
}

exports.detail = function(req, res, next) {
	res.json(req.todo);
};

exports.update = function(req, res, next) {
	Todo.findByIdAndUpdate(req.params.id, req.body, function(err, post) {
		if (err) return next(err);
		res.json(post);
	});
};

exports.remove = function(req, res, next) {
	Todo.findByIdAndRemove(req.params.id, req.body, function(err, post) {
		if (err) return next(err);
		res.json(post);
	});
};
